src/V3SchedReplicate.cpp

```diff
@@ -0,0 +1,256 @@
+    const VNUser1InUse user1InUse;
+    const auto getVarVertex = [&](AstVarScope* vscp) {
+        if (!vscp->user1p()) vscp->user1p(new VarVertex{graphp, vscp});
+        return vscp->user1u().to<VarVertex*>();
+    };
+
+    const auto addEdge = [&](Vertex* fromp, Vertex* top) {
+        new V3GraphEdge{graphp, fromp, top, 1};
+    };
+
+    // Predicate for whether a read of the given variable triggers this block
+    std::function<bool(AstVarScope*)> readTriggersThisLogic;
+
+    const auto addLogic = [&](Region region, AstScope* scopep, AstActive* activep) {
+        AstSenTree* const senTreep = activep->sensesp();
+
+        std::function<bool(AstVarScope*)> readTriggersThisLogic;
+
+        const VNUser4InUse user4InUse;
+
+        if (senTreep->hasClocked()) {
+            // Clocked logic is never triggered by reads
+            readTriggersThisLogic = [](AstVarScope*) { return false; };
+        } else if (senTreep->hasCombo()) {
+            // Combinational logic is always triggered by reads
+            readTriggersThisLogic = [](AstVarScope*) { return true; };
+        } else {
+            UASSERT_OBJ(senTreep->hasHybrid(), activep, "unexpected");
+            // Hybrid logic is triggered by all reads, except for reads of the explicit
+            // sensitivities
+            readTriggersThisLogic = [](AstVarScope* vscp) { return !vscp->user4(); };
+            senTreep->foreach<AstVarRef>([](const AstVarRef* refp) {  //
+                refp->varScopep()->user4(true);
+            });
+        }
+
+        for (AstNode* nodep = activep->stmtsp(); nodep; nodep = nodep->nextp()) {
+            LogicVertex* const lvtxp = new LogicVertex{graphp, scopep, senTreep, nodep, region};
+            const VNUser2InUse user2InUse;
+            const VNUser3InUse user3InUse;
+
+            nodep->foreach<AstVarRef>([&](AstVarRef* refp) {
+                AstVarScope* const vscp = refp->varScopep();
+                VarVertex* const vvtxp = getVarVertex(vscp);
+
+                // If read, add var -> logic edge FIXMEV5: explain the third term in the condition
+                if (refp->access().isReadOrRW() && !vscp->user3SetOnce()
+                    && readTriggersThisLogic(vscp) && !vscp->user2()) {  //
+                    addEdge(vvtxp, lvtxp);
+                }
+                // If written, add logic -> var edge
+                if (refp->access().isWriteOrRW() && !vscp->user2SetOnce()) {  //
+                    addEdge(lvtxp, vvtxp);
+                }
+            });
+        }
+    };
+
+    logicRegions.m_pre.foreach ([&](AstScope* scopep, AstActive* activep) {  //
+        addLogic(ACTIVE, scopep, activep);
+    });
+    logicRegions.m_active.foreach ([&](AstScope* scopep, AstActive* activep) {  //
+        addLogic(ACTIVE, scopep, activep);
+    });
+    logicRegions.m_nba.foreach ([&](AstScope* scopep, AstActive* activep) {  //
+        addLogic(NBA, scopep, activep);
+    });
+
+    return std::unique_ptr<Graph>{graphp};
```

Comment:
The graphp can be constructed as unique instead.

Reply:
FWIW, a unique_ptr can't be declared const (movable things generally cannot), so you loose the fact that graphp cannot be re-bound in this function, but I'm ok with that.